### PR TITLE
Add create consumer

### PR DESCRIPTION
**Pull request by Jarema**

### CODE REVIEW

File: async-nats/src/jetstream/mod.rs [hunk 1]
```diff
@@ -13,6 +13,7 @@
 
 use crate::Client;
 
+pub mod consumer;
```
Comment (caspervonb) [inline async-nats/src/jetstream/mod.rs:16]: Think you forgot to git add something 🙃

File: async-nats/src/jetstream/consumer.rs [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    /// Allows for a variety of options that determine how this consumer will receive messages
+    pub deliver_policy: DeliverPolicy,
+    /// Used in combination with `DeliverPolicy::ByStartSeq` to only select messages arriving
+    /// after this sequence number.
+    #[serde(
+        default,
+        skip_serializing_if = "Option::is_none",
+        rename = "opt_start_seq"
+    )]
+    pub opt_start_sequence: Option<u64>,
```
Comment (caspervonb) [inline async-nats/src/jetstream/consumer.rs:154]: We should roll this into the `DeliverPolicy` enum, I need a refresher on how we can do that with serde but it is possible.

```suggestion
    pub opt_start_sequence: Option<u64>,
```

Can follow up with it.

File: async-nats/src/jetstream/consumer.rs [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    /// Used in combination with `DeliverPolicy::ByStartTime` to only select messages arriving
+    /// after this time.
+    #[serde(default, skip_serializing_if = "is_default", with = "rfc3339::option")]
+    pub opt_start_time: Option<DateTime>,
```
Comment (caspervonb) [inline async-nats/src/jetstream/consumer.rs:158]: As above, roll into the almighty enum.

```suggestion
    pub opt_start_time: Option<DateTime>,
```

File: async-nats/src/jetstream/consumer.rs [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+    #[serde(rename = "last")]
+    Last = 1,
+    /// New will only deliver new messages that are received by the `JetStream` server
+    /// after the consumer is created.
+    #[serde(rename = "new")]
+    New = 2,
+    /// `ByStartSeq` will look for a defined starting sequence to the consumer's configured `opt_start_seq`
+    /// parameter.
+    #[serde(rename = "by_start_sequence")]
+    ByStartSeq = 3,
```
Comment (caspervonb) [inline async-nats/src/jetstream/consumer.rs:267]:
```suggestion
    ByStartSequence = 3,
```

File: async-nats/src/jetstream/consumer.rs [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    fn default() -> ReplayPolicy {
+        ReplayPolicy::Instant
+    }
+}
+
+fn is_default<T: Default + Eq>(t: &T) -> bool {
+    t == &T::default()
+}
+/// A UTC time
+pub type DateTime = time::OffsetDateTime;
```
Comment (caspervonb) [inline async-nats/src/jetstream/consumer.rs:332]: Do we need to alias this? Doesn't have any type parameters it simplifies.

```suggestion
```

**Review by caspervonb — CHANGES_REQUESTED** [review]
Bit over dead code,  and probably some naming top bikeshed over.
Actual consumer creation lgtm tho.

File: async-nats/src/jetstream/consumer.rs [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+#[derive(Debug, Default, Serialize, Deserialize, Clone, Copy, PartialEq, Eq)]
+pub struct SequencePair {
+    /// How far along the consumer has progressed
+    pub consumer_seq: u64,
+    /// The aggregate for all stream consumers
+    pub stream_seq: u64,
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub(crate) struct CreateConsumerRequest {
```
[the remark this answers is not in the document]
Comment (Jarema) [reply async-nats/src/jetstream/consumer.rs:70]: Hm, I liked typed request, but it seems this one is really used only once, even in legacy codebase.

File: async-nats/src/jetstream/stream.rs [hunk 7]
```diff
@@ -10,13 +10,55 @@
...
     pub info: StreamInfo,
+    pub(crate) prefix: String,
+    pub(crate) context: Context,
+}
+
+impl Stream {
+    pub async fn create_consumer<C: Into<ConsumerConfig>>(
+        &self,
+        config: C,
+    ) -> Result<ConsumerInfo, Error> {
```
[the remark this answers is not in the document]
Comment (Jarema) [reply async-nats/src/jetstream/stream.rs:36]: Imo create returns ConsumerInfo, get_or_create returns Consumer - so its consistent with streams API.

**Review by caspervonb — APPROVED** [review]
lgtm

**Pull request merged**